// clarity-angular/popover/dropdown/providers/dropdown.service.js
/*
 * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable, Optional, SkipSelf } from "@angular/core";
import { Subject } from "rxjs/Subject";
var RootDropdownService = (function () {
    function RootDropdownService() {
        this._changes = new Subject();
    }
    Object.defineProperty(RootDropdownService.prototype, "changes", {
        get: function () {
            return this._changes.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    RootDropdownService.prototype.closeMenus = function () {
        this._changes.next(false);
    };
    return RootDropdownService;
}());
export { RootDropdownService };
RootDropdownService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
RootDropdownService.ctorParameters = function () { return []; };
export function clrRootDropdownFactory(existing) {
    return existing || new RootDropdownService();
}
export var ROOT_DROPDOWN_PROVIDER = {
    provide: RootDropdownService,
    useFactory: clrRootDropdownFactory,
    deps: [[new Optional(), new SkipSelf(), RootDropdownService]]
};
