// clarity-angular/popover/dropdown/dropdown-trigger.js
/*
 * Copyright (c) 2016 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, HostListener } from "@angular/core";
import { IfOpenService } from "../../utils/conditional/if-open.service";
import { Dropdown } from "./dropdown";
var DropdownTrigger = (function () {
    function DropdownTrigger(dropdown, ifOpenService) {
        this.dropdown = dropdown;
        this.ifOpenService = ifOpenService;
        this.isRootLevelToggle = true;
        // if the containing dropdown has a parent, then this is not the root level one
        if (dropdown.parent) {
            this.isRootLevelToggle = false;
        }
    }
    Object.defineProperty(DropdownTrigger.prototype, "active", {
        get: function () {
            return this.ifOpenService.open;
        },
        enumerable: true,
        configurable: true
    });
    DropdownTrigger.prototype.onDropdownTriggerClick = function (event) {
        this.ifOpenService.toggleWithEvent(event);
    };
    return DropdownTrigger;
}());
export { DropdownTrigger };
DropdownTrigger.decorators = [
    { type: Directive, args: [{
                // We support both selectors for legacy reasons
                selector: "[clrDropdownTrigger],[clrDropdownToggle]",
                host: {
                    "[class.dropdown-toggle]": "isRootLevelToggle",
                    "[class.dropdown-item]": "!isRootLevelToggle",
                    "[class.expandable]": "!isRootLevelToggle",
                    "[class.active]": "active"
                }
            },] },
];
/** @nocollapse */
DropdownTrigger.ctorParameters = function () { return [
    { type: Dropdown, },
    { type: IfOpenService, },
]; };
DropdownTrigger.propDecorators = {
    'onDropdownTriggerClick': [{ type: HostListener, args: ["click", ["$event"],] },],
};
