// clarity-angular/popover/signpost/signpost.js
/*
 * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, ContentChild, } from "@angular/core";
import { IfOpenService } from "../../utils/conditional/if-open.service";
import { SignpostTriggerDirective } from "./signpost-trigger.directive";
/*********
 *
 * @class Signpost
 *
 * @description
 * Class used to configure and control the state of a Signpost and its associated SignpostContent.
 * It supports the clrPosition with a 'right-middle' default.
 *
 */
var Signpost = (function () {
    function Signpost() {
        /**********
         * @property useCustomTrigger
         *
         * @description
         * Flag used to determine if we need to use the default trigger or a user supplied trigger element.
         *
         * @type {boolean}
         */
        this.useCustomTrigger = false;
    }
    Object.defineProperty(Signpost.prototype, "customTrigger", {
        /**********
         * @property signPostTrigger
         *
         * @description
         * Uses ContentChild to check for a user supplied element with the SignpostTriggerDirective on it.
         *
         * @type {SignpostTriggerDirective}
         */
        set: function (trigger) {
            this.useCustomTrigger = !!trigger;
        },
        enumerable: true,
        configurable: true
    });
    return Signpost;
}());
export { Signpost };
Signpost.decorators = [
    { type: Component, args: [{
                selector: "clr-signpost",
                template: "\n        <ng-container *ngIf=\"!useCustomTrigger\">\n            <button\n                type=\"button\"\n                class=\"signpost-action btn btn-small btn-link\"\n                clrSignpostTrigger>\n                <clr-icon shape=\"info\"></clr-icon>\n            </button>\n        </ng-container>\n        \n        <ng-content></ng-content>\n    ",
                host: { "[class.signpost]": "true" },
                providers: [IfOpenService]
            },] },
];
/** @nocollapse */
Signpost.ctorParameters = function () { return []; };
Signpost.propDecorators = {
    'customTrigger': [{ type: ContentChild, args: [SignpostTriggerDirective,] },],
};
